content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst

```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
 
 .. include:: ../../global.txt
 
+Interacting with hardware devices is one of the more frequent activities
+in embedded systems programming. It is also one of the most enjoyable
+because you can make something happen in the physical world. There's a
+reason that making an LED blink is the "hello world" of embedded
+programming. Not only is it easy to do, it is surprisingly satisfying. I
+suspect that even the developers of "Full Authority Digital Engine
+Controllers" (FADEC) |mdash| the computers that are in complete, total
+control of commercial airline engines |mdash| have fond memories of
+making an LED blink early in their careers. And of course a blinking LED
+is a good way to indicate application status, especially if off-board
+I/O is limited, which is often the case.
+
+Working at the device register level can be error prone and relatively
+slow, in terms of source-lines-of-code (SLOC) produced. That's partly
+because the hardware is in some cases complicated, and partly because of
+the way the software is written. Using bit masks for setting and
+clearing bits is not a readable approach, comparatively speaking.
+There's just not enough information transmitted to the reader. It might
+be clear enough when written, but will you see it that way months later?
+Readability is important because programs are read many more times than
+they are written. Also, an unreadable program is more difficult to
+maintain, and maintenance is where most money is spent in long-lived
+applications. Comments can help, until they are out of date. Then they
+are an active hindrance.
+
+For example, what do you think the following code does? This is real
+code, where :c:`temp` and :c:`temp2` are unsigned 32-bit integers:
+
+.. code-block:: c
+
+   temp = ((uint32_t)(GPIO_AF) << ((uint32_t)((uint32_t)GPIO_PinSource & (uint32_t)0x07) * 4)) ;
+   GPIOx->AFR[GPIO_PinSource >> 0x03] &= ~((uint32_t)0xF << ((uint32_t)((uint32_t)GPIO_PinSource & (uint32_t)0x07) * 4)) ;
+   temp_2 = GPIOx->AFR[GPIO_PinSource >> 0x03] | temp;
+   GPIOx->AFR[GPIO_PinSource >> 0x03] = temp_2;
+
+That's unfair to ask, absent any context. The code configures a general
+purpose I/O (GPIO) pin on an Arm microcontroller for one of the
```

Comment:
Minor editorial question: isn't the uppercase abbreviation ("ARM") more commonly used than "Arm"?

Reply:
IIUC "Arm" is the newer form, as it is no longer an acronym. To be checked.

Reply:
Confirmed "Arm" is the current casing.